### PR TITLE
Update badge rarity calculation

### LINKED ISSUE #193: Store rarities as a separate column in the `BADGES` relation
Store rarities as a separate column in the `BADGES` relation

Extended from @abompard's advice here https://github.com/fedora-infra/tahrir/pull/704#discussion_r2321424835

### LINKED ISSUE #280: Compute badge rarities on the database API layer
Compute badge rarities on the database API layer

This extends the feature introduced as a part of https://github.com/fedora-infra/tahrir/pull/704. We need to introduce a column in the `badges` relation that stores the current rarity of the badge (i.e. `X`, `S`, `A`, `B`, `C`, `D`). This will be based on a calculation on the basis of the badge's ownership as done in the aforementioned pull request changes.

As to when it should be done, I like to think that we can perhaps do it whenever someone gets awarded a badge and since this specific implementation is lower level as compared to the superficial one mentioned above, this should not be that computationally expensive - but hey, I am also open to other opinions too.

Alternatively, we can have the rarities regenerated whenever a new badge gets introduced. The rarities would refresh less frequently but this also means that this is computationally cheaper. We could also do this once per release cycle but that might be too slow and we would have to find ways to track the release dates too.

Thoughts? Opinions?

### PR DESCRIPTION
**Pull request by Daniel-1600**
closes #280

### CODE REVIEW

File: tahrir_api/dbapi.py [hunk 1]
```diff
@@ -57,12 +58,9 @@ def __init__(self, dburi=None, session=None, autocommit=True, notification_callb
 
     def team_exists(self, team_id):
         """
-        Check to see if this team already exists in the database
 
-        :type team_id: str
         :param team_id: The ID of a Team
         """
```
Comment (gridhead) [inline tahrir_api/dbapi.py:64]: Do you mind explaining why you removed these comments?

I do hope they were not inconveniencing you in any manner.

File: tahrir_api/dbapi.py [hunk 2]
```diff
@@ -384,6 +382,85 @@ def get_all_badges(self):
 
         return self.session.query(Badge)
 
+    def compute_badge_rarities(self):
```
Comment (gridhead) [inline tahrir_api/dbapi.py:388]: You seem to have forgotten to include the `@autocommit` decorator here.

Or was it intentional by any chance?

File: tahrir_api/migrations/versions/5b1e956d70b5_add_rarities_table_and_rarity_id_to_.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
+"""add rarities table and rarity_id to badges
+
+Revision ID: 5b1e956d70b5
+Revises: 459d4cc47d13
```
Comment (gridhead) [inline tahrir_api/migrations/versions/5b1e956d70b5_add_rarities_table_and_rarity_id_to_.py]: Should this not be `caeca369a786` since that is where we introduced the `legacy` column?

Let me fix that for you.

File: tahrir_api/migrations/versions/5b1e956d70b5_add_rarities_table.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+Create Date: 2026-05-21 16:57:28.719850
+"""
+
+import sqlalchemy as sa
+from alembic import op
+
+
+# revision identifiers, used by Alembic.
+revision = "5b1e956d70b5"
+down_revision = "caeca369a786"
```
Comment (gridhead) [inline tahrir_api/migrations/versions/5b1e956d70b5_add_rarities_table.py:14]: Ok, now I am genuinely confused.

You did refer to `caeca369a786` here so this revision is aware about that change.

**Review by gridhead — CHANGES_REQUESTED** [review]
I have included some changes from my end.

But do pull/rebase first before making changes.

Also, please honestly answer the next question.

Have you used AI tooling here in this PR or not?

File: tahrir_api/migrations/versions/5b1e956d70b5_add_rarities_table.py (hunk 4 above)
Comment (Daniel-1600) [reply tahrir_api/migrations/versions/5b1e956d70b5_add_rarities_table.py:14]: I had started working on this issue before legacy column was merged  the so the revision I was using was for 459d4cc47d13 , I can see I corrected one and left the other

File: tahrir_api/dbapi.py (hunk 1 above)
Comment (Daniel-1600) [reply tahrir_api/dbapi.py:64]: Sorry about that, must have been accidentally removed during conflict resolution. I'll restore them.

**Comment by Daniel-1600** [comment]
> I have included some changes from my end.
> 
> But do pull/rebase first before making changes.
> 
> Also, please honestly answer the next question.
> 
> Have you used AI tooling here in this PR or not?

Yes, I used AI tooling to assist with parts of this PR, particularly for debugging and working through the rebase. The implementation decisions and code are mine, but I had AI assistance along the way.

**Comment by gridhead** [comment]
If you have had AI assistance along the way, why did you not say so in the commit messages?

We have referenced the [AI Assisted Contribution Policy](https://docs.fedoraproject.org/en-US/council/policy/ai-contribution-policy/) document countless times in the past.

Frankly, at this point in time, it is getting rather frustrating having to specify it again and again.

Especially so when this feels like a futile practice because you just keep ignoring what we say.

We never expected polished changes and we do assist with reviews/suggestions, do we not?

See, you are not utilizing (not one, not two, but) three mentors - and that is entirely on you.

File: tahrir_api/dbapi.py (hunk 1 above)
Comment (gridhead) [reply tahrir_api/dbapi.py:64]: I did that already.

Focus on other changes.

File: tahrir_api/migrations/versions/5b1e956d70b5_add_rarities_table.py (hunk 4 above)
Comment (gridhead) [reply tahrir_api/migrations/versions/5b1e956d70b5_add_rarities_table.py:14]: As long as you keep that in mind.

That is all we really ask.

**Comment by Daniel-1600** [comment]
Hey @gridhead , after pulling your latest changes and running tox, the format check fails with 21 files needing reformatting. But when I run black directly it says all files are unchanged.

**Comment by Daniel-1600** [comment]
> If you have had AI assistance along the way, why did you not say so in the commit messages?
> 
> We have referenced the [AI Assisted Contribution Policy](https://docs.fedoraproject.org/en-US/council/policy/ai-contribution-policy/) document countless times in the past.
> 
> Frankly, at this point in time, it is getting rather frustrating having to specify it again and again.
> 
> Especially so when this feels like a futile practice because you just keep ignoring what we say.
> 
> We never expected polished changes and we do assist with reviews/suggestions, do we not?
> 
> See, you are not utilizing (not one, not two, but) three mentors - and that is entirely on you.

I take full ownership of that mistake.I also should’ve leaned more on the mentors instead of handling it this way.I’ll do better going forward and be transparent next time.

**Comment by Daniel-1600** [comment]
@gridhead I updated the comments on ai tooling and the tests they are still failing  this is what I tried so far https://github.com/fedora-infra/tahrir-api/pull/285#issuecomment-4533286440

File: tahrir_api/dbapi.py [hunk 5]
```diff
@@ -888,7 +969,7 @@ def get_all_issuers(self):
 
         return self.session.query(Issuer)
 
-    def get_all_assertions(self, begin: Optional[int] = None, limit: Optional[int] = None):
+    def get_all_assertions(self, begin: int | None = None, limit: int | None = None):
```
Comment (sdglitched) [inline tahrir_api/dbapi.py:973]: `\` is not supported in Python 3.9 and was introduced in 3.10. I believe the py39 environment can be removed from the test.

File: tests/test_dbapi.py [hunk 6]
```diff
@@ -850,3 +859,167 @@ def test_badge_legacy_in_as_dict(api, dummy_badge_id, legacy_value, expected):
...
+        if i < 5:
+            api.add_assertion(badge_id, f"zero{i}@test.com", None, f"link_{i}")
+
+    api.compute_badge_rarities()
+    badges = api.get_all_badges().all()
+    assert all(b.rarity_id is not None for b in badges)
+    # The badge with 0 assertions should be in X tier (least owned)
+    zero_badge = next(b for b in badges if b.name == "ZeroBadge_5")
+    x_tier = api.session.query(Rarity).filter(Rarity.name == "X").first()
+    assert zero_badge.rarity_id == x_tier.id
```
Comment (sdglitched) [inline tests/test_dbapi.py:991]: @Daniel-1600, the current tests are too much fragmented, please consolidate them.

For example, below conditions can be checked in be a single test thereby reducing the multiple fixture calls:
- Check all 6 tiers created with correct names
- Check all badges assigned
- Check equal distribution
- Check limits are valid
- Check rarest badge  (X tier)
- Check commonest badge (D tier)

Remaining edge cases can have different test cases.

**Review by sdglitched — CHANGES_REQUESTED** [review]
@Daniel-1600, please check my comment about the test cases and implement them.

@gridhead, is it possible to force run the tests for other environment instead of the py39 environment?

File: tahrir_api/dbapi.py [hunk 7]
```diff
@@ -384,6 +384,87 @@ def get_all_badges(self):
...
+            .all()
+        )
+
+        userpoll = self.session.query(Person).count()
+
+        if userpoll == 0 or not counts:
+            return
+
+        # build the same accodict structure as the frontend script
+        accodict = OrderedDict()
```
Comment (gridhead) [inline tahrir_api/dbapi.py]: Does it have to be an `OrderedDict()`?

You are using `sorted()` anyway later.

File: tahrir_api/dbapi.py [hunk 8]
```diff
@@ -384,6 +384,87 @@ def get_all_badges(self):
...
+            # update the rarities table row for this tier
+            rarity_row = self.session.query(Rarity).filter(Rarity.name == rare_name).first()
+            if rarity_row:
+                rarity_row.lower_limit = lower
+                rarity_row.upper_limit = upper
+            else:
+                rarity_row = Rarity(name=rare_name, lower_limit=lower, upper_limit=upper)
+                self.session.add(rarity_row)
+
+            self.session.flush()
```
Comment (gridhead) [inline tahrir_api/dbapi.py]: `flush` once after all `rarity` entries are created.

Then assign `rarity_id` on the consequent pass.

File: tahrir_api/migrations/versions/5b1e956d70b5_add_rarities_table.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+        sa.Column("name", sa.Unicode(8), nullable=False),
+        sa.Column("lower_limit", sa.Float(), nullable=False),
+        sa.Column("upper_limit", sa.Float(), nullable=False),
+        sa.PrimaryKeyConstraint("id"),
+        sa.UniqueConstraint("name"),
+    )
+    op.add_column(
+        "badges",
+        sa.Column("rarity_id", sa.Integer(), sa.ForeignKey("rarities.id"), nullable=True),
+    )
```
Comment (gridhead) [inline tahrir_api/migrations/versions/5b1e956d70b5_add_rarities_table.py:32]: Consider creating `sa.Index("ix_badges_rarity_id", "rarity_id")`.

Even though it would be a relatively small table - it feels complete.

File: poetry.lock
Comment (gridhead) [inline poetry.lock:1]: Did you change some dependencies in between?

If not, please remove this from your PR changes.

File: tahrir_api/model.py [hunk 10]
```diff
@@ -59,6 +59,8 @@ class Badge(DeclarativeBase):
     created_on = Column(DateTime, nullable=False, default=datetime.datetime.now)
     tags = Column(Unicode(128))
     legacy = Column(Boolean, default=False, nullable=False, server_default=false())
+    rarity_id = Column(Integer, ForeignKey("rarities.id"), nullable=True)
+    rarity = relationship("Rarity")
```
Comment (gridhead) [inline tahrir_api/model.py]: Should you not have `back_populates` here to speed queries?

Like "show all badges with rarity X" or something like that?

File: tahrir_api/model.py [hunk 11]
```diff
@@ -312,3 +314,15 @@ def __pygments__(self):
             pygments.formatters.HtmlFormatter(**html_args),
         ).strip()
         return html
+
+
+class Rarity(DeclarativeBase):
```
Comment (gridhead) [inline tahrir_api/model.py:320]: Where is the `as_dict()` serialization method?

It is there in other relation specifications too.

**Review by gridhead — CHANGES_REQUESTED** [review]
Please follow the suggested changes here.

File: tahrir_api/dbapi.py [hunk 12]
```diff
@@ -384,6 +384,87 @@ def get_all_badges(self):
...
+                Assertion.badge_id,
+                func.count(Assertion.id).label("count"),
+            )
+            .group_by(Assertion.badge_id)
+            .all()
+        )
+
+        userpoll = self.session.query(Person).count()
+
+        if userpoll == 0 or not counts:
```
Comment (gridhead) [inline tahrir_api/dbapi.py]:
```suggestion
        if userpoll == 0:
```

File: tahrir_api/model.py (hunk 11 above)
Comment (gridhead) [inline tahrir_api/model.py:320]: Something like this is in order.

```python
def as_dict(self):
    return dict(
        name=self.name,
        lower_limit=self.lower_limit,
        upper_limit=self.upper_limit,
    )
```

File: tahrir_api/model.py
Comment (gridhead) [inline tahrir_api/model.py:84]: Why is there no reference to `rarity` here?

Try something like this.

```python
def as_dict(self):
    if self.image.startswith("http"):
        image = self.image
    else:
        image = "/pngs/" + self.image
    return dict(
        version="0.5.0",
        name=self.name,
        image=image,
        description=self.description,
        criteria=self.criteria,
        issuer=self.issuer.as_dict(),
        created_on=time.mktime(self.created_on.timetuple()),
        tags=self.tags,
        legacy=self.legacy,
        rarity=self.rarity.as_dict() if self.rarity else None,
    )
```

File: tahrir_api/utils.py [hunk 13]
```diff
@@ -32,7 +32,7 @@ def convert_name_to_id(name):
 
     badge_id = name.lower().replace(" ", "-")
     bad = ['"', "'", "(", ")", "*", "&", "?"]
-    replacements = dict(zip(bad, [""] * len(bad)))
+    replacements = dict(zip(bad, [""] * len(bad), strict=False))
```
Comment (gridhead) [inline tahrir_api/utils.py:35]: Good enough.

We do not support Python 3.9 anyway anymore.

**Review by gridhead — CHANGES_REQUESTED** [review]
Might I please ask you to ensure the _completeness_ of your changes before you send your PR in?

They do not have to be perfect but at least they should pass the minimum criteria for being evaluated.

**Review by gridhead — APPROVED** [review]
LGTM 🌮
